### DISCUSSION TITLE
Generating conditions with multiple tables dynamically

### DISCUSSION THREAD

**Discussion by coudy** [linq2db/linq2db#3197] [Q&A] [answered]
I currently have a query like this:

```
from a in Database.A
from b in Database.B
  .Where(...)
from c in Database.C
  .Where(...)
  .DefaultIfEmpty()
where searchTerm == null ||
      Sql.Ext.SqlServer().Contains(searchTerm, a) ||
      Sql.Ext.SqlServer().Contains(searchTerm, b) ||
      Sql.Ext.SqlServer().Contains(searchTerm, c)
```

I want to generalise this to multiple search terms, i.e.:

```
WHERE
  (CONTAINS(A.*, @searchTerm1) OR CONTAINS(B.*, @searchTerm1) OR CONTAINS(C.*, @searchTerm1)) AND
  (CONTAINS(A.*, @searchTerm2) OR CONTAINS(B.*, @searchTerm2) OR CONTAINS(C.*, @searchTerm2)) AND
  (CONTAINS(A.*, @searchTerm3) OR CONTAINS(B.*, @searchTerm3) OR CONTAINS(C.*, @searchTerm3))...
```

I generally understand how to work with `Expression<...>` (and LINQKit) but I'm not sure how to create condition expressions for my query above and bind them to the appropriate elements.

I could try this in raw method chain syntax but I have 10 tables in the actual query, so at that point it's probably easier to abandon LINQ and hand-craft the query.

Is this something that can currently be done with linq2db?

Alternatively, is it possible to write where clauses in raw SQL with linq2db?

**Comment by coudy** [comment]
Does anyone have an answer? I'd really appreciate any pointers.

**Comment by artelk** [comment]
Maybe something like
```cs
var searchTerms = new [] { "aaa", "bbb", ... };

//...
where searchTerms.All(searchTerm =>
               searchTerm == null ||
                 Sql.Ext.SqlServer().Contains(searchTerm, a) ||
                 Sql.Ext.SqlServer().Contains(searchTerm, b) ||
                 Sql.Ext.SqlServer().Contains(searchTerm, c))
//...
```

**Reply by coudy** [reply]
This would be a really good solution if implemented in linq2db but it currently fails with:  `LinqToDB.Linq.LinqException: Sequence '....searchTerms' cannot be converted to SQL.`

**Reply by sdanyliv** [reply]
Well, latest version 3.4.5 supports such syntax. Enjoy.

**Comment by sdanyliv** [comment, marked as answer]
Missed this discussion. No way to do that in Query variant.
For method chain, I have answered similar question: 
SO https://stackoverflow.com/a/67666993/10646316
```cs
var query =
    from a in Database.A
    from b in Database.B
        .Where(...)
    from c in Database.C
        .Where(...)
        .DefaultIfEmpty()
    select new
    {
        a, b, c, d...
    }

query = query.FilterByItems(searchTerms, (e, s) => Sql.Ext.SqlServer().Contains(s, e.a, e.b, e.c, e.d, ...), true);
```

**Reply by coudy** [reply]
Thanks. This is essentially what I resorted to, albeit using LINQKit instead of an extension method. The main issue for me is that I have to include entire rows in the `select`, which massively increases the size of the result set.

Would you consider implementing something closer to @artelk's answer?

**Reply by sdanyliv** [reply]
> The main issue for me is that I have to include entire rows in the select, which massively increases the size of the result set.

It's not true. You can make final projection
```cs
var result = query.Select(x => x.a.SomeField);
```
And only one  field will be retrieved.

> Would you consider implementing something closer to @artelk's answer?

Several times ;) I have to find time.

**Reply by coudy** [reply]
Thanks for clarifying, didn't realise that'd work!

I might try to get some of our devs involved. No promises, though. :-)

**Reply by artelk** [reply]
There is a big enough gun for this: ExpressionMethodAttribute and Compile() magic 😄 
```cs
// use this All2 instead of All in the query
[ExpressionMethod(nameof(All2Impl))]
public static bool All2<T>(this IEnumerable<T> source, Func<T, bool> predicate) => source.All(predicate);

private static Expression<Func<IEnumerable<T>, Func<T, bool>, bool>> All2Impl<T>() =>
    (source, predicate) => All2Impl(source).Compile()(predicate);

private static Expression<Func<Func<T, bool>, bool>> All2Impl<T>(IEnumerable<T> source)
{
    var list = source.ToList();
    if (list.Count == 0)
        return predicate => true;

    var predicate = Expression.Parameter(typeof(Func<T, bool>));
    Expression body = null;
    foreach (var item in list)
    {
        var invoke = Expression.Invoke(predicate, Expression.Constant(item));
        body = body == null ? invoke : Expression.AndAlso(body, invoke);
    }
    return Expression.Lambda<Func<Func<T, bool>, bool>>(body, predicate);
}
```
(didn't properly test this)

**Reply by sdanyliv** [reply]
Looks good. I've never tested such approach ;)
